### PR TITLE
feat(apigatewayv2): create http apis from an OpenAPI specification

### LINKED ISSUE #19190: (module name): short issue description
### Description

Adding OpenAPI support for HTTP APIs (API Gateway V2).

Reference: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-open-api.html

### Use Case

It would be great to develop both REST and HTTP API using OpenAPI files (one file per API gateway, of course).

### Proposed Solution

Import the OpenAPI specification per the reference above.

### Other information

_No response_

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by miguel-a-calles-mba**
Adding OpenAPI support for HTTP APIs (API Gateway V2).

Closes #19190
Based on #19641

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `cdk-integ` to deploy the infrastructure and generate the snapshot (i.e. `cdk-integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by ncaq** [comment]
When I tried to import it and use it in my product, I realized that unlike Rest, HttpApi performs a lot of processing such as defaultIntegration and default stage in its own constructor, so if I could not use it, I felt it would be difficult to use it in the same way. I felt that it would be difficult to use the HttpApiBase in the same way.
I think it is necessary to move those processes to HttpApiBase or to be able to specify ApiDefinition in HttpAPI Prop.

**Comment by miguel-a-calles-mba** [comment]
@otaviomacedo , where do I add the documentation?

File: packages/@aws-cdk/aws-apigatewayv2/README.md [hunk 1]
```diff
@@ -447,3 +448,33 @@ const webSocketApi = new apigwv2.WebSocketApi(this, 'mywsapi',{
...
+## OpenAPI Definition
+
+CDK supports creating an HTTP API by importing an OpenAPI definition file. It currently supports OpenAPI v2.0 and OpenAPI
+v3.0 definition files. Read more about [Configuring a REST API using
+OpenAPI](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-import-api.html).
+
+The following code creates an HTTP API using an external OpenAPI definition JSON file -
+
+```ts
+declare const integration: apigateway.Integration;
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/README.md:460]: Seems like something is missing here. You're not actually using `integration` in this readme and it's not set to any value here.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import * as cxapi from '@aws-cdk/cx-api';
+
+import { Node } from 'constructs';
+import { CfnApi } from '../apigatewayv2.generated';
+import { IHttpApi } from './api';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts:12]: This is no longer necessary.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
+import * as s3 from '@aws-cdk/aws-s3';
+import * as s3_assets from '@aws-cdk/aws-s3-assets';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts:4]: This import doesn't need to be separate.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Thank you for your contribution! We can't provide a meaningful review until the build succeeds and we have integration tests, but I have a few initial comments inline.

**Comment by TheRealAmazonKendra** [comment]
> If the OpenAPI title and the apiName in the props did not match, the following error occurred.
> 
> ```
> Redundant fields [name="foo-deve"] provided when either Body or BodyS3Location is not empty
> ```
> 
> This is non-intuitive and it would be better to remove the apiName from the props.

I agree with this point. Let's create `HttpApiPropsBase` with the props that actually need to be in both and extend that instead. Then edit `HttpApiProps` to extend `HttpApiPropsBase` with the other props that aren't needed in `SpecHttpApiProps`.

File: packages/@aws-cdk/aws-apigatewayv2/README.md (hunk 1 above)
Comment (miguel-a-calles-mba) [reply packages/@aws-cdk/aws-apigatewayv2/README.md]: I updated the code to address this comment.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts (hunk 2 above)
Comment (miguel-a-calles-mba) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts]: I updated the code to address this comment.

**Comment by miguel-a-calles-mba** [comment]
I updated the code to address the `HttpApiPropsBase`, `HttpApiProps` and `SpecHttpApiProps` changes.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  /**
+   * Called when the specification is initialized to allow this object to bind
+   * to the stack, add resources and have fun.
+   *
+   * @param scope The binding scope. Don't be smart about trying to down-cast or
+   * assume it's initialized. You may just use it as a construct scope.
+   */
+  public abstract bind(scope: Construct): ApiDefinitionConfig;
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts:41]: Based off the context I'm seeing here, it looks like this is a function the end user isn't meant to use. Given that, let's mark it internal and name it `_bind` instead.

**Comment by TheRealAmazonKendra** [comment]
We still need [integration tests](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) for this change, but I'll provide a round of feedback to what's been provided so far to keep this moving.

File: packages/@aws-cdk/aws-apigatewayv2/test/http/api.test.ts [hunk 5]
```diff
@@ -531,6 +532,63 @@ describe('HttpApi', () => {
   });
 });
 
+describe('SpecHttpApi', () => {
+  test('default', () => {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/test/http/api.test.ts:536]: I'd like to see more unit test cases overall. Looks like there aren't any for `api-definition`.

File: packages/@aws-cdk/aws-apigatewayv2/test/http/api.test.ts [hunk 6]
```diff
@@ -531,6 +532,63 @@ describe('HttpApi', () => {
...
+      apiDefinition: ApiDefinition.fromInline(oas),
+    });
+
+    Template.fromStack(stack).hasResourceProperties('AWS::ApiGatewayV2::Api', {
+      Body: oas,
+    });
+
+    expect(api.apiEndpoint).toBeDefined();
+    expect(api.apiId).toBeDefined();
+    expect(api.httpApiId).toBeDefined();
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-apigatewayv2/test/http/api.test.ts:588]: These should be tested against the expected output, not just that they're defined.

**Review by TheRealAmazonKendra — DISMISSED** [review]

**Review by TheRealAmazonKendra — DISMISSED** [review]
Putting back into changes requested to reflect status.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+ */
+export class S3ApiDefinition extends ApiDefinition {
+  private bucketName: string;
+
+  constructor(bucket: s3.IBucket, private key: string, private objectVersion?: string) {
+    super();
+
+    if (!bucket.bucketName) {
+      throw new Error('bucketName is undefined for the provided bucket');
+    }
```
[the remark this answers is not in the document]
Comment (miguel-a-calles-mba) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api-definition.ts:98]: I used the code from https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-apigateway/lib/api-definition.ts#L139 as a starting point.

I'll update the code there too.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Perhaps this piece is still in progress, but the integration test will need to have the snapshot output added as well. See our [Integration Test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) documentation. Looks like you may be still using the old style of test as well.

**Comment by TheRealAmazonKendra** [comment]
BTW, I know I have a LOT of feedback and changes requested so I want to also take the time to thank you for all your work on this. I'm confident we can refine this into an excellent user experience!

**Comment by miguel-a-calles-mba** [comment]
> BTW, I know I have a LOT of feedback and changes requested so I want to also take the time to thank you for all your work on this. I'm confident we can refine this into an excellent user experience!

No worries, @TheRealAmazonKendra. I'm glad somebody is looking at this PR. I have been waiting for a long while for a review since my first version of this PR #19641.

I'll do my best to address the comments.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Putting this back in requested changes to reflect the status.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Putting this back in requested changes to reflect the status.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be closed. Please create a new PR for these changes if you think this decision has been made in error.

**Comment by miguel-a-calles-mba** [comment]
@TheRealAmazonKendra, sorry, I have been supporting other priorities and this took a lower priority.

**Comment by adminy** [comment]
Great, so this never made it into cdk then ... 😢

**Pull request closed without merging**